### DISCUSSION TITLE
ID3v2 metadata in HLS streams

### DISCUSSION THREAD

**Discussion by fkane** [savonet/liquidsoap#3152] [Q&A]
Hi, I'm trying to integrate my station's HLS stream into Sonos devices. In order to display metadata such as artist, title, and cover art, they require it to be embedded in ID3v2 frames, and have specific requirements on how to embed URL's to the album art.

Is there any existing mechanism for embedding ID3v2 metadata into HLS streams, and any existing hooks I can use to implement the WXXX tag for a URL to the album art in the way they want?

Their requirements are [here](https://developer.sonos.com/build/content-service-add-features/supported-audio-formats/), and pasted below for convenience:

For HLS content, use [ID3v2 frames](http://id3.org/id3v2.3.0#Declared_ID3v2_frames) for metadata, as defined in the [ID3 tag version 2.3.0 documentation](http://id3.org/id3v2.3.0). Sonos players read the following ID3v2 tags:

COMM—comments
PRIV—private frame
TALB—album/move/show
TCMP—iTunes compilation
TCOM—composer
TCON—content type
TIT2—title/song name/content
TPE1—lead performer, soloist
TPE2—band/orchestra
TPOS—part of a set
TRCK—track number
WXXX—user defined url link

Use the WXXX tag if you want your stream to have artwork that changes, like an Internet radio stream that changes artwork for each song. We use this tag to specify the size of the artwork and the URL. To do this, you must separate these strings with a null character, such as \0. The first string should start with “artworkURL_” followed by the width of the image in pixels and “x”. For example, this string will be set to “artworkURL_640x” for an image that is 640 pixels wide. The second string should be the full album art URI (example: http://server/path/art.jpg”). Together, the strings would look like this: “artworkURL_640x\0http://example.com/artwork/1341414.jpg”.

**Comment by toots** [comment]
Let's add this, it's been a very much expected feature for a while. I've started the work to support rendering custom id3v2 frames: https://github.com/savonet/ocaml-metadata/pull/1

**Reply by fkane** [reply]
Thanks so much for tackling this. How would you go about using this new functionality? Would I just use existing metadata functions (such as the annotate protocol) using the tag names as the key values? Or is it more directly tied to the HLS output?

**Reply by toots** [reply]
It should be automatic based on the normal metadata flow. I'll have a test build for you today.

**Reply by fkane** [reply]
Hi, just want to make sure you're not waiting on me to test something... I see there is an hls-id3v2 branch now; should I be working with that?

**Reply by toots** [reply]
Hi! Tests would be great but I am slowly completing the PR. I pushed a bugfix upstream to FFmpeg, added a test. Next documentation and PR description. Then tests, yes! Particularly, I'm enabling ID3 by default and I would like to make sure that this does not cause regressions with existing HLS streams.
